src/sentry/workflow_engine/migration_helpers/alert_rule.py

```diff
@@ -447,30 +447,60 @@ def create_workflow(
     )
 
 
-def create_detector(
+def get_detector_field_values(
     alert_rule: AlertRule,
-    project_id: int,
     data_condition_group: DataConditionGroup,
+    project_id: int | None = None,
     user: RpcUser | None = None,
-) -> Detector:
-    return Detector.objects.create(
-        project_id=project_id,
-        enabled=True,
-        created_by_id=user.id if user else None,
-        name=alert_rule.name,
-        workflow_condition_group=data_condition_group,
-        type=MetricAlertFire.slug,
-        description=alert_rule.description,
-        owner_user_id=alert_rule.user_id,
-        owner_team=alert_rule.team,
-        config={
+) -> dict[str, Any]:
+    detector_field_values = {
+        "name": alert_rule.name,
+        "description": alert_rule.description,
+        "workflow_condition_group": data_condition_group,
+        "owner_user_id": alert_rule.user_id,
+        "owner_team": alert_rule.team,
+        "config": {
             "threshold_period": alert_rule.threshold_period,
             "sensitivity": alert_rule.sensitivity,
             "seasonality": alert_rule.seasonality,
             "comparison_delta": alert_rule.comparison_delta,
             "detection_type": alert_rule.detection_type,
         },
+    }
+    if project_id is not None:
+        # these fields are only set on create, not update
+        detector_field_values.update(
+            {
+                "project_id": project_id,
+                "enabled": True,
+                "created_by_id": user.id if user else None,
+                "type": MetricAlertFire.slug,
+            }
+        )
+    return detector_field_values
+
+
+def create_detector(
+    alert_rule: AlertRule,
+    project_id: int,
+    data_condition_group: DataConditionGroup,
+    user: RpcUser | None = None,
+) -> Detector:
+    detector_field_values = get_detector_field_values(
+        alert_rule, data_condition_group, project_id, user
     )
+    return Detector.objects.create(**detector_field_values)
+
+
+def update_detector(
+    alert_rule: AlertRule,
+    detector: Detector,
+):
+    if detector.workflow_condition_group is None:
+        raise MissingDataConditionGroup
+    detector_field_values = get_detector_field_values(alert_rule, detector.workflow_condition_group)
+    detector.update(**detector_field_values)
```

Comment:
🤔 i wonder if we should use the `DetectorValidator` to handle this for us. There is already an update method on it and it could handle creation too. that would allow us to validate the `detector_field_values` as well.

Reply:
I don't disagree but I looked into it a bit and I'd want to refactor `get_detector_validator` a bit to be reusable and then update this function to be like:

```
def update_detector(
    alert_rule: AlertRule,
    detector: Detector,
) -> Detector:
    if detector.workflow_condition_group is None:
        raise MissingDataConditionGroup

    validator = get_detector_validator(project, "metric_alert_fire", detector_field_values, instance=detector)
    if not validator.is_valid():
        # TODO make this more descriptive
        raise Exception()

    validator.save()
    return detector
```

which I think is better addressed in a follow up PR 👍